build-support/bin/generate_completions.py

```diff
@@ -0,0 +1,164 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import json
+import subprocess
+
+BASH_COMPLETION_TEMPLATE = """
+# DO NOT EDIT.
+# This script is autogenerated by build-support/bin/generate_autocomplete.py
+
+function pants_completions()
+{{
+    readonly PANTS_GOALS="\\
+        {goals}"
+
+    {goal_options}
+
+    local current_word previous_word previous_goal
+
+    current_word=${{COMP_WORDS[COMP_CWORD]}}
+    previous_word=${{COMP_WORDS[COMP_CWORD-1]}}
+    previous_goal=$(get_previous_goal)
+
+    case "$previous_goal" in
+    {goal_cases}
+    esac
+
+    # Show goals, unless the tab completion has a hyphen - then show global/goal options instead
```

Comment:
Will this also show scoped subsystem options?

Reply:
We can, with approximately no effort. Other than `changed` (which I have) but it came down the fact that I'd just never used any subsystems from the command line, no other reason.

Reply:
Yeah, maybe this should be optional. 95% of the time subsystem options are set in config and not on the cli.

Ooooh, maybe we've finally found a sensible use for `advanced=True` options: Only non-advanced options are in the cli completion! (or at least by default)

Reply:
> Gotcha. I'm happy to merge this as-is, so we can play around with it internally, under the assumption that we'll need to do more work to make it usable at large.

Before merging, I'd like to put the content of the auto-completions to bed, so that any future PRs are more about Pants integration, rather than integration AND content. 

I can add all subsystems as well, so we have a picture of what a "full" auto-complete could look like - and in the integration, maybe we can pare down what should be in the `advanced` or wherever else.

Reply:
Just a quick test on mainline Pants, global options drops from 627 to 251 when skipping `advanced` help.

Reply:
Added subsystem options